### PR TITLE
Add support for WSL with HaneWIN NFS server

### PR DESCRIPTION
**Pull request by DzeryCZ**
This pull request is bringing support for Windows subsystem for Linux (WSL) with [HaneWIN NFS server](https://www.hanewin.net/nfs-e.htm).

I've tested these changes on Mac as well, everything works as expected.

### CODE REVIEW

File: docker-machine-nfs.sh [hunk 1]
```diff
@@ -135,7 +135,7 @@ setPropDefaults()
   prop_machine_name=
   prop_shared_folders=()
   prop_nfs_config="-alldirs -mapall="$(id -u):$(id -g)
-  prop_mount_options="noacl,async,nfsvers=3"
+  prop_mount_options="noacl,async,vers=3"
```
Comment (tonivdv) [inline docker-machine-nfs.sh]: why did you change this?

File: docker-machine-nfs.sh [hunk 2]
```diff
@@ -91,21 +91,21 @@ EOF
 # @args:    error-message
 echoError ()
 {
-  echo "\033[0;31mFAIL\n\n$1 \033[0m"
+  printf "\033[0;31mFAIL\n\n$1 \033[0m\n"
```
Comment (tonivdv) [inline docker-machine-nfs.sh:94]: Those changes (and other similar changes) are not really related to the support for HaneWIN ... I prefer to keep the PRs related to one topic and not mix things. Another PR could be created with only semantic changes .

I'd also be interested why this change would be required ... ?!

File: docker-machine-nfs.sh [hunk 3]
```diff
@@ -326,7 +326,7 @@ lookupMandatoryProperties ()
   fi
 
   prop_nfshost_ip=$(VBoxManage list hostonlyifs |
-    grep "${prop_network_id}$" -A 3 | grep IPAddress |
+    grep "${prop_network_id}" -A 3 | grep IPAddress |
```
Comment (tonivdv) [inline docker-machine-nfs.sh]: can you elaborate this change?

File: docker-machine-nfs.sh [hunk 4]
```diff
@@ -398,7 +424,8 @@ configureBoot2Docker()
   # (this will override an existing /var/lib/boot2docker/bootlocal.sh)
 
   local bootlocalsh='#!/bin/sh
-  sudo umount /Users'
+  sudo umount /Users
+  sudo umount /c/Users'
```
Comment (tonivdv) [inline docker-machine-nfs.sh]: Won't this cause warnings if the volume does not exist? Maybe better to create a bootlocal depending on the OS?

File: docker-machine-nfs.sh (hunk 2 above)
Comment (DzeryCZ) [reply docker-machine-nfs.sh:94]: At some places you're using `printf` and at some `echo`. 

Firstly I wanted to unify it, and secondly on WSL and on Linux bash `echo` doesn't print symbols with backslash correctly. Instead of plain `echo` you have to use either `echo -e` or `printf`.

Example:
```
jerry@JerryPC /c/Users/Jerry/projects/docker-machine-nfs (support-for-wsl) $ echo "\033[0;31mTest\n\n$1 \033[0m"                                                                                                                
\033[0;31mTest\n\n \033[0m 
jerry@JerryPC /c/Users/Jerry/projects/docker-machine-nfs (support-for-wsl) $ echo -e "\033[0;31mTest\n\n$1 \033[0m"                                                                                                             
Test 
     
     
jerry@JerryPC /c/Users/Jerry/projects/docker-machine-nfs (support-for-wsl) $ printf "\033[0;31mTest\n\n$1 \033[0m\n"
Test 
     
     
jerry@JerryPC /c/Users/Jerry/projects/docker-machine-nfs (support-for-wsl) $  
```

I'm taking those changes back and will include in another PR.

File: docker-machine-nfs.sh (hunk 3 above)
Comment (DzeryCZ) [reply docker-machine-nfs.sh]: Because it unfortunately doesn't find any result.

Please see:

- Without `$`:
```
jerry@JerryPC /c/Users/Jerry/projects/docker-machine-nfs (support-for-wsl) $ VBoxManage list hostonlyifs | grep "VirtualBox Host-Only Ethernet Adapter #7"                                                                 
VBoxNetworkName: HostInterfaceNetworking-VirtualBox Host-Only Ethernet Adapter #7
```
- With `$`
```
jerry@JerryPC /c/Users/Jerry/projects/docker-machine-nfs (support-for-wsl) $ VBoxManage list hostonlyifs | grep "VirtualBox Host-Only Ethernet Adapter #7$" 
```
*No ouptupt*

It's because of difference of line endings in Linux (LF) and Windows (CRLF) systems. Used `VBoxManage` is a windows binary and outputs CRLF.

---

There are two ways how to handle this: 
- Either do not look for line ends - omit `$` at the end 
*or*
- Remove `\r` (CR) in console output - e.g.: `VBoxManage list hostonlyifs | tr -d '\r' | grep "VirtualBox Host-Only Ethernet Adapter #7$"`

I like the first solution more, because it seems to be a bit cleaner. What do you think?

Comment (DzeryCZ) [reply docker-machine-nfs.sh]: At the end the removing of `\r` has to be there, because it causes an issue later on.

Comment (tonivdv) [reply docker-machine-nfs.sh]: The problem is that removing the $ will regress a previous bug => https://github.com/adlogix/docker-machine-nfs/commit/ef23e63644113f05c8f4e9532e704a4d466a3f46

Comment (tonivdv) [reply docker-machine-nfs.sh]: More info on the bug => https://github.com/adlogix/docker-machine-nfs/issues/71

Comment (DzeryCZ) [reply docker-machine-nfs.sh]: I see.  So I've put the trailing `$` back and I'm removing `\r` from console output.

```
prop_nfshost_ip=$(VBoxManage list hostonlyifs | tr -d '\r' |
    grep "${prop_network_id}$" -A 3 | grep IPAddress |
    cut -d ':' -f2 | xargs);
```

Comment (tonivdv) [reply docker-machine-nfs.sh]: It would be cool if anybody with a mac / linux can test this as I/we are not using the library anymore :/

File: docker-machine-nfs.sh [hunk 5]
```diff
@@ -418,8 +457,8 @@ configureBoot2Docker()
   local file="/var/lib/boot2docker/bootlocal.sh"
 
   docker-machine ssh $prop_machine_name \
-    "echo '$bootlocalsh' | sudo tee $file && sudo chmod +x $file && sync" > /dev/null
-
+    "echo '$bootlocalsh' | sudo tee $file && sudo chmod +x $file && sync" < /dev/null > /dev/null
```
Comment (tonivdv) [inline docker-machine-nfs.sh:472]: Can you elaborate this change too? You made that change in different places ... wondering what the reason is, since this part does work quite well ..

Comment (DzeryCZ) [reply docker-machine-nfs.sh:472]: `docker-machine` on windows waits for keyboard input. So this script hangs until you press Enter. `< /dev/null` prevents this behavioral.

I've tested it also on Mac and it works without any issue.

**Comment by DzeryCZ** [comment]
Hello,

Could somebody please take a look at this updated PR?

Many thanks,
Jerry

File: docker-machine-nfs.sh (hunk 1 above)
Comment (tonivdv) [reply docker-machine-nfs.sh]: Think you closed this by mistake?!

**Review by tonivdv — COMMENTED** [review]
> Hello,
> 
> Could somebody please take a look at this updated PR?
> 
> Many thanks,
> Jerry

I was waiting for any other feedback from others. But seems there isn't. I got one small unanswered question and after that I guess it's fine.

File: docker-machine-nfs.sh (hunk 1 above)
Comment (DzeryCZ) [reply docker-machine-nfs.sh]: It seems, that `nfsvers=` and `vers=` has the same [behavioural](http://manpages.ubuntu.com/manpages/bionic/man5/nfs.5.html). 

Will remove the change anyway,

**Comment by tonivdv** [comment]
@DzeryCZ thanks and sorry for the delay
